### ISSUE TITLE
new command: driftctl export

### ISSUE THREAD

**Issue by sjourdan** [cloudskiff/driftctl#933]
**Description**

As a driftctl cli user
I want to export my driftctl result
So that I can use it in another program

**Example**

TBD but an idea: 

```
$ driftctl export --format json
[...]
{

}
```

**Comment by sundowndev** [comment]
It's unclear to me what does the export command would perform under the hoods. Does it need to perform an actual scan ?

> default driftctl mode doesn't include resource details (experimental --deep mode does)

We could make deep mode being a requirement to use the export feature.

> there's no need for any processing middleware there

It's still possible to create the export before running any middleware.

**Comment by sjourdan** [comment]
> It's unclear to me what does the export command would perform under the hoods. Does it need to perform an actual scan ?

Yes, `driftctl export` should scan the resources like the real scan, but also add resources content from the tf provider without any processing (we don't care about it there). Like a "raw" _deep mode_ by default if you prefer. 

> We could make deep mode being a requirement to use the export feature.

True, `driftctl export` can't work without either the real deep mode or an equivalent!

> It's still possible to create the export before running any middleware.

interesting, do you mean a prototype version of the `driftctl export` command can be done to reuse the current tfplan PoC and make it less complicated to maintain in the long run?

**Comment by sundowndev** [comment]
> interesting, do you mean a prototype version of the driftctl export command can be done to reuse the current tfplan PoC and make it less complicated to maintain in the long run?

I think having to create a new command to make another kind of scan will be hard to maintain. So may be we could embed the export feature in a simple `--export` flag for the existing scan command ? In this way, we don't rewrite the scan command and still be able to avoid useless processing. Unlike the abandoned PoC was suggesting, the export format shouldn't be considered as a scan output.

Another option: we don't perform a scan in the export command but take result of a scan as argument, does that make sense to you ?

**Comment by wbeuil** [comment]
The scan in its root feature (read state + scan remote + first part of the analysis) is indeed what we need to do here.

*BUT* what we meant by "processing" is all the attributes that we decided voluntarily to remove for the sake of driftctl (e.g. look at `initAwsCloudfrontDistributionMetaData`) and all the middlewares that could potentially remove other essential attributes.

I don't think we could do that in just one command, thus creating another command for it.

**Comment by sundowndev** [comment]
> all the attributes that we decided voluntarily to remove for the sake of driftctl

@wbeuil You're right I didn't think about metadata we remove. I better understand the choice of abandoning the previous PoC.

**Comment by eliecharra** [comment]
Wow, it's quite a big one RFC for sure.

First thing, resource anonymization is another topic, and for me, it should remain in a dedicated RFC. We're gonna to end with a big mess in our brain if we try to solve multiples problems in a single RFC.

While reading I don't get what we really want here. We talk everywhere about "result"
> I want to export my result to the Terraform JSON plan format

Maybe the first thing can be to explain what we want exactly to export ?

> there's no need for any processing middleware there

Why ? One more time, what do we expect as a result ? If this is actual `scan` result, it does not make sense not to run middleware on a scan result. Middleware are here to ensure consistency between IaC and real life resources, if you simply not run them you'll end with an inconsistent output 🤔 

> Yes, driftctl export should scan the resources like the real scan, but also add resources content from the tf provider without any processing (we don't care about it there). Like a "raw" deep mode by default if you prefer.

`but also add resources content from the tf provider` I don't get it, what is a tf provider for you ?

**TL;DR**
I'm stopping the review here since I don't get what we really want here.
I think we should start from a clear explanation of what we want, with real life usecase.

> So that I can use it somewhere else

"somewhere else" sounds very unclear. This should be clarified.

Implementation details will came after a deep reflection about what we want, and then we'll talk about UX/UI and implementation details.

**Comment by moadibfr** [comment]
I have to agree with Elie on the export part.
I'll have to add that I'm not sure the anonymization part is the same thing as the export part. For debug purpose we need a full scan run and most of the think we want to anonymize are in the logs. I'm pretty sure these are modification of scan and not a new command.
